x-pack/plugins/data_enhanced/server/search/session/session_service.ts

```diff
@@ -25,21 +33,150 @@ import {
+      namespaces: ['*'],
+    });
+    this.logger.debug(`getAllMappedSavedObjects | Got ${res.saved_objects.length} items`);
+    return res.saved_objects;
+  }
+
+  private clearSessions = () => {
+    const curTime = moment();
+
+    this.sessionSearchMap.forEach((sessionInfo, sessionId) => {
+      if (
+        moment.duration(curTime.diff(sessionInfo.insertTime)).asSeconds() >
+        INMEM_TRACKING_TIMEOUT_SEC
+      ) {
+        this.logger.debug(`clearSessions | Deleting expired session ${sessionId}`);
+        this.sessionSearchMap.delete(sessionId);
+      } else if (sessionInfo.retryCount >= MAX_UPDATE_RETRIES) {
+        this.logger.warn(`clearSessions | Deleting failed session ${sessionId}`);
+        this.sessionSearchMap.delete(sessionId);
+      }
+    });
+  };
+
+  public async monitorMappedIds() {
+    try {
+      this.clearSessions();
+
+      if (!this.sessionSearchMap.size) return;
+      this.logger.debug(`monitorMappedIds | Map contains ${this.sessionSearchMap.size} items`);
+
+      const savedSessions = await this.getAllMappedSavedObjects();
+      const updatedSessions = await this.updateAllSavedObjects(savedSessions);
+
+      updatedSessions.forEach((updatedSavedObject) => {
+        const sessionInfo = this.sessionSearchMap.get(updatedSavedObject.id)!;
+        if (updatedSavedObject.error) {
+          // Retry next time
+          sessionInfo.retryCount++;
+        } else if (updatedSavedObject.attributes.idMapping) {
+          // Delete the ids that we just saved, avoiding a potential new ids being lost.
+          Object.keys(updatedSavedObject.attributes.idMapping).forEach((key) => {
+            sessionInfo.ids.delete(key);
+          });
+          // If the session object is empty, delete it as well
+          if (!sessionInfo.ids.entries.length) {
+            this.sessionSearchMap.delete(updatedSavedObject.id);
+          } else {
+            sessionInfo.retryCount = 0;
+          }
+        }
+      });
+    } catch (e) {
+      this.logger.error(`monitorMappedIds | Error fetching sessions. ${e}`);
+    }
+  }
+
+  private async updateAllSavedObjects(
+    activeMappingObjects: Array<SavedObject<BackgroundSessionSavedObjectAttributes>>
+  ) {
+    if (!activeMappingObjects.length) return [];
+
+    this.logger.debug(`updateAllSavedObjects | Updating ${activeMappingObjects.length} items`);
+    const updatedSessions = activeMappingObjects
+      ?.filter((so) => !so.error)
+      .map((sessionSavedObject) => {
+        const sessionInfo = this.sessionSearchMap.get(sessionSavedObject.id);
+        const idMapping = sessionInfo ? Object.fromEntries(sessionInfo.ids.entries()) : {};
+        sessionSavedObject.attributes.idMapping = {
+          ...sessionSavedObject.attributes.idMapping,
+          ...idMapping,
+        };
+        return sessionSavedObject;
+      });
+
+    const updateResults = await this.internalSavedObjectsClient.bulkUpdate<BackgroundSessionSavedObjectAttributes>(
```

Comment:
I wonder if this could throw a conflict error and if we should immediately retry if it did. (I am not sure when SO api throws conflict error, just've seen 409 error before)

Reply:
As far as I understand, `bulkUpdate` returns an error per object. You can see those processed in `monitorMappedIds`, when checking each response item for errors.